internal/resources/cluster/data_source_cluster.go

```diff
@@ -25,116 +25,162 @@ import (
+}
+
+func withTerraformConstructType(t terraformConstruct) terraformOption {
+	return func(config *terraformConfig) {
+		config.terraformConstructType = t
+	}
+}
 
-		return false, nil
+func withOperationType(o operationType) terraformOption {
+	return func(config *terraformConfig) {
+		config.operationType = o
 	}
+}
+
+type readContextType func(ctx context.Context, d *schema.ResourceData, m interface{}) (diags diag.Diagnostics)
 
-	timeoutValueData, _ := d.Get(waitKey).(string)
+func dataSourceTMCClusterRead(opts ...terraformOption) readContextType {
+	cfg := &terraformConfig{}
 
-	if ctx.Value(contextMethodKey{}) != "create" {
-		timeoutValueData = "do_not_retry"
+	for _, o := range opts {
+		o(cfg)
 	}
 
-	switch timeoutValueData {
-	case "do_not_retry":
-		_, err = getClusterResourceRetryableFn()
-	case "":
-		fallthrough
-	case "default":
-		_, attachClusterWithKubeconfig := d.GetOk(attachClusterKey)
-		fn := constructFullname(d)
+	return func(ctx context.Context, d *schema.ResourceData, m interface{}) (diags diag.Diagnostics) {
+		config := m.(authctx.TanzuContext)
 
-		if fn.ManagementClusterName == attachedValue && !attachClusterWithKubeconfig {
-			_, err = getClusterResourceRetryableFn()
-			break
+		// Warning or errors can be collected in a slice type
+		var (
+			resp *clustermodel.VmwareTanzuManageV1alpha1ClusterGetClusterResponse
+			err  error
+		)
+
+		getClusterResourceRetryableFn := func() (retry bool, err error) {
+			resp, err = config.TMCConnection.ClusterResourceService.ManageV1alpha1ClusterResourceServiceGet(constructFullname(d))
+			if err != nil {
+				if clienterrors.IsNotFoundError(err) {
+					_ = schema.RemoveFromState(d, m)
+
+					switch cfg.terraformConstructType {
+					case constructTypeResource:
+						return false, nil
+
+					case constructTypeDataSource:
+						return false, errors.Wrapf(err, "Tanzu Mission Control cluster entry not found, name : %s", d.Get(NameKey))
+					}
+				}
+
+				return true, errors.Wrapf(err, "Unable to get Tanzu Mission Control cluster entry, name : %s", d.Get(NameKey))
+			}
+
+			d.SetId(resp.Cluster.Meta.UID)
+
+			if !strings.EqualFold(string(clustermodel.VmwareTanzuManageV1alpha1ClusterPhaseREADY), string(*resp.Cluster.Status.Phase)) {
+				log.Printf("[DEBUG] waiting for cluster(%s) to be in %v phase, present phase:%v", constructFullname(d).ToString(), clustermodel.VmwareTanzuManageV1alpha1ClusterPhaseREADY, *resp.Cluster.Status.Phase)
+				return true, nil
+			}
+
+			return false, nil
 		}
 
-		timeoutValueData = "3m"
+		timeoutValueData, _ := d.Get(waitKey).(string)
 
-		fallthrough
-	default:
-		timeoutDuration, parseErr := time.ParseDuration(timeoutValueData)
-		if parseErr != nil {
-			log.Printf("[INFO] unable to prase the duration value for the key %s. Defaulting to 3 minutes(3m)"+
-				" Please refer to 'https://pkg.go.dev/time#ParseDuration' for providing the right value", waitKey)
+		if cfg.operationType != operationCreate {
```

Comment:
all the context related logic should be written at one place for ease of reading /